Document/0x05e-Testing-Cryptography.md

```diff
@@ -1,16 +1,70 @@
 ## Android Cryptographic APIs
 
+### Overview
+
 In the chapter "[Cryptography for Mobile Apps](0x04g-Testing-Cryptography.md)", we introduced general cryptography best practices and described typical flaws that can occur when cryptography is used incorrectly in mobile apps. In this chapter, we'll go into more detail on Android's cryptography APIs. We'll show how to identify uses of those APIs in the source code and how to interpret the configuration. When reviewing code, make sure to compare the cryptographic parameters used with the current best practices linked from this guide.
 
-### Testing the Configuration of Cryptographic Standard Algorithms (MSTG-CRYPTO-2, MSTG-CRYPTO-3 and MSTG-CRYPTO-4)
+We can identify key components of cryptography system in Android:
 
-#### Overview
+- [Provider](0x05e-Testing-Cryptography.md#provider)
+- KeyStore - see the section [KeyStore](0x05d-Testing-Data-Storage.md#keystore) in the chapter "Testing Data Storage"
+- KeyChain - see the section [KeyChain](0x05d-Testing-Data-Storage.md#keychain) in the chapter "Testing Data Storage"
 
 Android cryptography APIs are based on the Java Cryptography Architecture (JCA). JCA separates the interfaces and implementation, making it possible to include several [security providers](https://developer.android.com/reference/java/security/Provider.html "Android Security Providers") that can implement sets of cryptographic algorithms. Most of the JCA interfaces and classes are defined in the `java.security.*` and `javax.crypto.*` packages. In addition, there are Android specific packages `android.security.*` and `android.security.keystore.*`.
 
+KeyStore and KeyChain provides APIs for storing and using keys (behind the scene, KeyChain API uses KeyStore system). These systems allow to administer the full lifecycle of the cryptographic keys. Requirements and guidance for implementation of cryptographic key management can be found in [Key Management Cheat Sheet](https://owasp.org/www-project-cheat-sheets/cheatsheets/Key_Management_Cheat_Sheet.html "Key Management Cheat Sheet"). We can identify following phases:
+
+- generating of a key
+- using of a key
+- storing of a key
+- archiving of a key
+- deleting of a key
+
+> Please note that storing of a key is analyzed in the chapter "[Testing Data Storage](0x05d-Testing-Data-Storage.md)".
+
+These phases are managed by Keystore/KeyChain system. However how the system works depends on how application developer used it. Then, you should focus on functions which are used by application developer. You should identify and verify following functions:
+
+- [Key generation](0x05e-Testing-Cryptography.md#key-generation)
+- [Random number generation](0x05e-Testing-Cryptography.md#random-number-generation)
+
+You can go further and verify how good is a key by performing [key attestation](0x05e-Testing-Cryptography.md#key-attestation).
```

Comment:
i thought key attestation was mostly about verifying the integrity of the Secure world in android. Not necessarily whether the key itself was ok? And is it already broadly available?

Reply:
My bad, I think about something else here (key lifecycle) and link it incorrectly.

Reply:
Additionally, I moved key attestation to Testing Data Storage as it is related closely to hardware-backed storage of keys.